actionpack/lib/action_dispatch/middleware/cookies.rb

```diff
@@ -249,39 +258,42 @@ def signed_or_encrypted
 
       private
 
-        def upgrade_legacy_signed_cookies?
+        def rotate_out_legacy_signed_cookies?
           request.secret_token.present? && request.secret_key_base.present?
         end
 
-        def upgrade_legacy_hmac_aes_cbc_cookies?
+        def rotate_out_legacy_aes_cbc_hmac_cookies?
           request.secret_key_base.present?                       &&
             request.authenticated_encrypted_cookie_salt.present? &&
             request.encrypted_signed_cookie_salt.present?        &&
             request.encrypted_cookie_salt.present?
         end
-    end
 
-    # Passing the ActiveSupport::MessageEncryptor::NullSerializer downstream
-    # to the Message{Encryptor,Verifier} allows us to handle the
-    # (de)serialization step within the cookie jar, which gives us the
-    # opportunity to detect and migrate legacy cookies.
-    module VerifyAndUpgradeLegacySignedMessage # :nodoc:
-      def initialize(*args)
-        super
-        @legacy_verifier = ActiveSupport::MessageVerifier.new(request.secret_token, serializer: ActiveSupport::MessageEncryptor::NullSerializer)
-      end
+        delegate :key_generator, :encrypted_cookie_rotate_out,
+          :signed_cookie_rotate_out, to: :request
 
-      def verify_and_upgrade_legacy_signed_message(name, signed_message)
-        deserialize(name, @legacy_verifier.verify(signed_message)).tap do |value|
-          self[name] = { value: value }
+        def encrypted_cookie_cipher
+          request.encrypted_cookie_cipher || "aes-256-gcm"
         end
-      rescue ActiveSupport::MessageVerifier::InvalidSignature
-        nil
-      end
 
-      private
-        def parse(name, signed_message)
-          super || verify_and_upgrade_legacy_signed_message(name, signed_message)
+        def authenticated_encrypted_cookie_salt
+          request.authenticated_encrypted_cookie_salt || ""
+        end
+
+        def encrypted_cookie_salt
+          request.encrypted_cookie_salt || ""
+        end
+
+        def encrypted_signed_cookie_salt
+          request.encrypted_signed_cookie_salt || ""
+        end
+
+        def signed_cookie_digest
+          request.signed_cookie_digest || "SHA1"
+        end
+
+        def signed_cookie_salt
+          request.signed_cookie_salt || ""
```

Comment:
Do we need to provide these empty string fallbacks?

Reply:
Don't think its necessary to have the fallbacks. I can change this.